File: solutions/clojure/say/2/src/say.clj
(ns say)

(def until-20
  (clojure.string/split
   "zero one two three four five six seven eight nine
    ten eleven twelve thirteen fourteen fifteen sixteen
    seventeen eighteen nineteen"
   #"\n? +"))

(def decenas
  (clojure.string/split
   "zero ten twenty thirty forty fifty sixty seventy eighty ninenty"
   #" "))

(def thousands
  (clojure.string/split
   "zero thousand million billion" #" "))

(defn quot-rem
  [n den]
  [(quot n den) (rem n den)])

(defn split-thousands
  [n]
  (->> [n 0]
       (iterate #(quot-rem (first %) 1000))
       (take-while #(not= % [0 0]))
       (drop 1)
       (map second)
       (map-indexed vector)))


(defn ensure-range
  [num]
  (when (neg? num) (throw (IllegalArgumentException. "Not valid negative number")))
  (when (< 999999999999 num) (throw (IllegalArgumentException. "Not valid for above 999999999999"))))

(defn ->words
  [[idx num]]
  (let [[h d] (quot-rem num 100)
        even? (zero? (rem d 10))]
    (clojure.string/trim
     (cond-> ""
       (pos? h)   (str (get until-20 h) " hundred")
       (< 0 d 20) (str " " (get until-20 d))
       (< 19 d)   (str " " (get decenas (quot d 10)))
       (and
        (< 19 d)
        (not even?)) (str "-" (get until-20 (rem d 10)))
       (and
        (pos? num)
        (pos? idx))  (str " " (get thousands idx))))))

(defn number
  [num]
  (ensure-range num)
  (cond
    (zero? num) "zero"
    (< num 20) (->words [0 num])
    :else      (->> (split-thousands num)
                    (map ->words)
                    reverse
                    (clojure.string/join " ")
                    clojure.string/trim)))

File: solutions/python/say/2/say.py
"""Convert integers into their English names."""

from collections.abc import Iterator

ONES = (
    "zero", "one", "two", "three", "four", "five", "six", "seven",
    "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen",
    "fifteen", "sixteen", "seventeen", "eighteen", "nineteen",
)
TENS = ("", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety")
SCALES = ("", "thousand", "million", "billion")
MAX_NUMBER = 999_999_999_999


def _say_under_thousand(number: int) -> str:
    if number < 20:
        return ONES[number]

    hundreds, remainder = divmod(number, 100)
    words = []
    if hundreds:
        words.extend((ONES[hundreds], "hundred"))
    if remainder:
        if remainder < 20:
            words.append(ONES[remainder])
        else:
            tens, ones = divmod(remainder, 10)
            words.append(TENS[tens] + (f"-{ONES[ones]}" if ones else ""))
    return " ".join(words)


def _chunks(number: int) -> list[int]:
    chunks = []

    while number:
        number, chunk = divmod(number, 1000)
        chunks.append(chunk)

    return chunks


def _numbered_chunks(number: int) -> Iterator[tuple[int, int]]:
    return reversed(list(enumerate(_chunks(number))))


def _say_chunk(chunk: int, scale: str) -> str:
    words = [word for word in (_say_under_thousand(chunk), scale) if word]
    return " ".join(words)


def say(number: int) -> str:
    """Return the English form for a number in the supported range."""
    if not 0 <= number <= MAX_NUMBER:
        raise ValueError("input out of range")

    if number == 0:
        return ONES[0]

    words = []
    for index, chunk in _numbered_chunks(number):
        if chunk:
            words.append(_say_chunk(chunk, SCALES[index]))

    return " ".join(words)
